src/shared/udev-util.c

```diff
@@ -91,3 +91,62 @@ int udev_parse_config_full(
 
         return 0;
 }
+
+static int device_monitor_handler(sd_device_monitor *monitor, sd_device *device, void *userdata) {
+        const char *sysname1 = userdata, *sysname2;
+
+        assert(device);
+        assert(sysname1);
+
+        if (sd_device_get_sysname(device, &sysname2) >= 0 && streq(sysname1, sysname2))
+                return sd_event_exit(sd_device_monitor_get_event(monitor), 0);
+
+        return 0;
+}
+
+int device_wait_for_initialization(sd_device *device, const char *subsystem) {
```

Comment:
But that's exactly what is happening. We "return the sd_device object when it was seen as initialized the first time". All we know is that udev thinks the object is initialized. We might get a state of the database from before the function call, or from after the function call, depending on timing, and the only thing we know that is that the state is from after initialization.

Reply:
device_wait_for_initialization() returns no sd_device in your version. It takes one on input, but none on output. Hence your version simply waits that the db contains an entry for the device, but the caller than continues with the non-initialized sd_device from earlier on anyway...

Reply:
sd_device is mutable. So it doesn't return anything, but the object that the caller holds a reference too becomes initalized.

Reply:
Hmm, it does? How is the data invalidated? Not sure I grok this?

Reply:
`sd_device_get_is_initialized` returns the value of `device->is_initialized`. It is a boolean that is set to true when the db entry under `/run/udev/data/` has been found for the device and successfully opened (and most likely read, barring some internal error). That's all "is_initalized" means.

Before `sd-device: attempt to read db again if it wasn't found`, it would be necessary to recreate the device object to recheck the presence of the database entry. But since that commit, `sd_device_get_is_initialized` will attempt to re-read the db entry if it wasn't found previously. 

I'll update the commit descriptions to include more information, hopefully this will clarify things.

Reply:
urks, i wasn't aware sd_device_get_is_initialized actually re-reads the db... it's a bit weird, that an accessor like this refreshes the data fully...